src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs

```diff
@@ -0,0 +1,204 @@
+            + SECTION_SIZE // type
+            + GetArraySectionSize(numberOfSections) // code
+            + SECTION_SIZE // data
+            + TERMINATOR_SIZE;
+        public static int GetArraySectionSize(int numberOfSections) => 3 + numberOfSections * 2;
+
+
+        private readonly ILogger? _logger;
+        private bool _loggerEnabled => _logger?.IsTrace ?? false;
+
+        public Eof1(ILogManager? logManager = null)
+        {
+            _logger = logManager?.GetClassLogger<Eof1>();
+        }
+
+        public bool ValidateCode(ReadOnlySpan<byte> container)
+        {
+            return TryParseEofHeader(container, out EofHeader? header) && ValidateBody(container, in header);
+        }
+
+        public bool TryParseEofHeader(ReadOnlySpan<byte> container, [NotNullWhen(true)] out EofHeader? header)
+        {
+            header = null;
+            if (!container.StartsWith(EOF_MAGIC))
+            {
+                if (_loggerEnabled)
+                    _logger.Trace($"EIP-3540 : Code doesn't start with Magic byte sequence expected {EOF_MAGIC.ToHexString(true)} ");
+                return false;
+            }
+            if (container[2] != VERSION)
+            {
+                if (_loggerEnabled)
+                    _logger.Trace($"EIP-3540 : Code is not Eof version {VERSION}");
+                return false;
+            }
+            if (container.Length < MINIMUM_HEADER_SIZE
+                + 1 + 1 + 2 // minimum type section body size
+                + 1) // minimum code section body size
+            {
+                if (_loggerEnabled)
+                    _logger.Trace($"EIP-3540 : Eof{VERSION}, Code is too small to be valid code");
+                return false;
+            }
+
+            ushort numberOfCodeSections = container[7..9].ReadEthUInt16();
+            int headerSize = CalculateHeaderSize(numberOfCodeSections);
+            int pos = 3;
+
+            if (container[pos] != KIND_TYPE)
+            {
+                if (_loggerEnabled)
+                    _logger.Trace($"EIP-3540 : Eof{VERSION}, Code header is not well formatted");
+                return false;
+            }
+            pos++;
+            SectionHeader typeSection = new()
+            {
+                Start = headerSize,
+                Size = container[pos..(pos + 2)].ReadEthUInt16()
+            };
+            pos += 2;
+
+            if (container[pos] != KIND_CODE)
+            {
+                if (_loggerEnabled)
+                    _logger.Trace($"EIP-3540 : Eof{VERSION}, Code header is not well formatted");
+                return false;
+            }
+            pos += 3; // kind_code(1) + num_code_sections(2)
+            List<SectionHeader> codeSections = new();
+            int lastEndOffset = typeSection.EndOffset;
+            for (ushort i = 0; i < numberOfCodeSections; i++)
+            {
+                SectionHeader codeSection = new()
```

Comment:
I think this should check if index is within bounds, in the case `0xef00010100010203000400...` the codeSection is truncated which makes the parsing treat DATA_KIND as the number of code section and that ends up going out of bounds here

Reply:
true. should we just try catch it?